posthog/models/cohort/sql.py

```diff
@@ -96,3 +96,10 @@
 WHERE team_id = %(team_id)s AND cohort_id = %(cohort_id)s
 GROUP BY person_id, cohort_id, team_id
 """
+
+CLEAR_STALE_COHORTPEOPLE = f"""
+INSERT INTO cohortpeople
+SELECT person_id, %(cohort_id)s AS cohort_id, %(team_id)s AS team_id, sign * -1, version AS version
+FROM cohortpeople
+WHERE team_id = %(team_id)s AND cohort_id = %(cohort_id)s AND sign > 0 AND version < %(version)s
```

Comment:
This filter on sign doesn't make too much sense to me. There are cohorts where they have a -ve sign, but no corresponding +ve sign, which is why they're not deleted.

See:

```
SELECT count(1), version FROM (
SELECT person_id, version
    FROM cohortpeople
    WHERE team_id = 2 AND cohort_id = 12863
    GROUP BY person_id, cohort_id, team_id, version
    HAVING sum(sign) < 0
    )
GROUP BY version
ORDER BY version desc
```

Reply:
I'm not _quite_ sure how only -ve signs were inserted, but probs not worth digging into this bug, given that we're changing the table schema soon

Reply:
That was the problem I was mentioning `The first “cancel” and the last “state” rows, if the number of “state” and “cancel” rows matches and the last row is a “state” row.` They're saved

https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/collapsingmergetree/

Reply:
Don't think I get what you're saying 🤔 .

Consider the case where we have more `cancel` than `state` rows (which I'm trying to show using the query above).

In this case, we need to insert a `state` row to delete this entry, but the current query doesn't do it, the current query only inserts `cancel` rows.

Reply:
If there's a cancel row _first_ and you try to add a state row after, these rows don't collapse because the state row comes after cancel.

Reply:
The collapsing depends on the ordering of the `state` and `cancel` rows where `state` needs to come before `cancel` for it to collapse. If `cancel` exists, inserting a `state` row doesn't collapse them

Reply:
Ah, I get it now, thanks 😅

Reply:
Hmm this is tricky. I don't think it's deterministic when trying to insert duplicate rows. Since our sort key includes all the columns, there's no way to insert multiple state rows uncollapsed (tested this locally)

Reply:
Huh, based on that, doesn't it imply that once you have an 'extra' cancel row, there's no way to get rid of it?

Reply:
that's what I'm feeling at the moment. I think this may be an edge case with how the sort key is set up on this table. Since the sort key occupies all the columns, you can't have more than 2 rows collapsing and as a result of the order of the rows is negative -> positive, there's no way to fully collapse.

Reply:
_sigh_. This is 💩 , since there's 100s of versions of each cohort with dangling -1s.

How do you feel about making this PR into a clean up task that uses async deletion code path that the new table will use? (but deleting from `cohortpeople` for now)

We can test it works in production well by deleting these bajillion cohort rows; thus getting confidence it will work well under load for the new table.

Reply:
yeah that's the only sensible option I can see at the moment too. Will give that a try here

Reply:
If we insert 2 state rows one by one, we might be able to resolve this 🤔 . But this is already too 💩 , I prefer the delete solution as well.